diff --git a/qjs.c b/qjs.c
index 4dd11f8..ae19184 100644
--- a/qjs.c
+++ b/qjs.c
@@ -41,6 +41,8 @@
 #include "cutils.h"
 #include "quickjs-libc.h"
 
+#include "path.h"
+
 extern const uint8_t qjsc_repl[];
 extern const uint32_t qjsc_repl_size;
 #ifdef CONFIG_BIGNUM
@@ -331,7 +333,9 @@ int main(int argc, char **argv)
         load_jscalc = !strcmp(exename, "qjscalc");
     }
 #endif
-    
+
+    set_exe_dir_from_exe_name(argv[0]);
+
     /* cannot use getopt because we want to pass the command line to
        the script */
     optind = 1;
